### ISSUE TITLE
Use "odd numbers" to represent integer values

### ISSUE THREAD

**Issue by j-hui** [ssm-lang/sslang#28]
We have settled on representing all values using (32-bit) words; these may either represent literal *values* where possible, as is the case with integers, enums, and other small primitve types, or *pointers* to heap-allocated objects.

The memory manager needs to distinguish between these types of values because it must be able to recursively drop pointers embedded in heap-allocated structures. Whether a heap-allocated structure contains pointers can depend on the variant (e.g., `Some bigObj : Maybe BigObj` contains a pointer to `BigObj` while `Nothing : Maybe BigObj` does not, so this determination must be made at runtime.

Since pointers will always point to word-aligned values in the heap, we are guaranteed that the LSB is 0 for pointers. We can safely represent up to 31 bits of information by value, as long as we encode them such that the LSB for these values is always 1. For example. to represent an integer value `x`, we encode it as `x << 1 | 1`. 

As a first step toward using words to represent a mix of values and pointers, we need to first adjust codegen to encode values using this scheme. It also needs to decode values back into integers to perform arithmetic before re-encoding them.

**Comment by EmilySillars** [comment]
Update on progress!
- I implemented marshaling/unmarshaling of SSLANG integers in Codegen.hs
and incorporated feedback from my draft PR.
- I added a regression test called 31bitInts.ssl

- Things I have tested and that work in 31bitInts.ssl:
integer literals
parameters of reference type
addition and subtraction of (1) and (2) together


- Things I would like to test:
Non-reference parameters to functions
Local variables of both reference and non-reference type
Bit Not
Multiplication and division

Errors I'm getting trying to implement these tests:

```
main ( cout : & Int, x : Int) -> () = //pattern match error
/*sslc: pattern should be desguared into pattern match
CallStack (from HasCallStack):
  error, called at src/IR/LowerAst.hs:176:18 in sslang-0.1.0.0-KI8f4RQ6phR54Leprkbw8R:IR.LowerAst*/

let x = (5:Int) //should already be desugared
/*sslc: Should already be desugared
CallStack (from HasCallStack):
  error, called at src/IR/LowerAst.hs:133:32 in sslang-0.1.0.0-KI8f4RQ6phR54Leprkbw8R:IR.LowerAst*/

after 10 , (cout : & Int) <- (~39):Int //syntax error
/*sslc: 6:33    [241+1] :Syntax error
CallStack (from HasCallStack):
  error, called at .stack-work/dist/x86_64-linux-tinfo6/Cabal-3.2.1.0/build/Front/Parser.hs:1199:5 in sslang-0.1.0.0-KI8f4RQ6phR54Leprkbw8R:Front.Parser*/

after 10 , (cout : & Int) <- (108*1):Int
/*sslc: Prelude.head: empty list*/
```

**Comment by j-hui** [comment]
Awesome!

For the errors you're getting:
- for passing multiple parameters, just pass them separately for now, i.e., `main (cout: &Int) (x: Int) -> ()`. We don't have structured binding just yet since we're blocked on pattern matching in the front end. 
- I'm not sure what the "should already be desugared" bug is. But you shouldn't need that type annotation. 
- we don't have any prefix operators right now other than `&` as a type operator. I'm not sure how best to fix that easily; you could add a `bnot` function and then fish for that in the lowering phase, but that might just add to bloat that shouldn't be there in the first place. 
- not quite sure what that's about either. Did you add in the typing rule for multiplication?

**Comment by j-hui** [comment]
Fixed by #41
